cluster-autoscaler/cloudprovider/openstack/openstack_nodegroup.go

```diff
@@ -0,0 +1,288 @@
+//
+// Represents a homogeneous collection of nodes within a cluster,
+// which can be dynamically resized between a minimum and maximum
+// number of nodes.
+type OpenstackNodeGroup struct {
+	openstackManager OpenstackManager
+	id               string
+
+	clusterUpdateMutex *sync.Mutex
+
+	minSize int
+	maxSize int
+	// Stored as a pointer so that when autoscaler copies the nodegroup it can still update the target size
+	targetSize *int
+
+	nodesToDelete      []*apiv1.Node
+	nodesToDeleteMutex sync.Mutex
+
+	waitTimeStep time.Duration
+}
+
+// WaitForClusterStatus checks once per second to see if the cluster has entered a given status.
+// Returns when the status is observed or the timeout is reached.
+func (ng *OpenstackNodeGroup) WaitForClusterStatus(status string, timeout time.Duration) error {
+	klog.V(2).Infof("Waiting for cluster %s status", status)
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(ng.waitTimeStep) {
+		clusterStatus, err := ng.openstackManager.GetClusterStatus()
+		if err != nil {
+			return fmt.Errorf("error waiting for %s status: %v", status, err)
+		}
+		if clusterStatus == status {
+			klog.V(0).Infof("Waited for cluster %s status", status)
+			return nil
+		}
+	}
+	return fmt.Errorf("timeout (%v) waiting for %s status", timeout, status)
+}
+
+// IncreaseSize increases the number of nodes by replacing the cluster's node_count.
+//
+// Takes precautions so that the cluster is not modified while in an UPDATE_IN_PROGRESS state.
+// Blocks until the cluster has reached UPDATE_COMPLETE.
+func (ng *OpenstackNodeGroup) IncreaseSize(delta int) error {
+	ng.clusterUpdateMutex.Lock()
+	defer ng.clusterUpdateMutex.Unlock()
+
+	if delta <= 0 {
+		return fmt.Errorf("size increase must be positive")
+	}
+
+	size, err := ng.openstackManager.NodeGroupSize(ng.id)
+	if err != nil {
+		return fmt.Errorf("could not check current nodegroup size: %v", err)
+	}
+	if size+delta > ng.MaxSize() {
+		return fmt.Errorf("size increase too large, desired:%d max:%d", size+delta, ng.MaxSize())
+	}
+
+	updatePossible, currentStatus, err := ng.openstackManager.CanUpdate()
+	if err != nil {
+		return fmt.Errorf("can not increase node count: %v", err)
+	}
+	if !updatePossible {
+		return fmt.Errorf("can not add nodes, cluster is in %s status", currentStatus)
+	}
+	klog.V(0).Infof("Increasing size by %d, %d->%d", delta, *ng.targetSize, *ng.targetSize+delta)
+	*ng.targetSize += delta
+
+	err = ng.openstackManager.UpdateNodeCount(ng.id, *ng.targetSize)
+	if err != nil {
+		return fmt.Errorf("could not increase cluster size: %v", err)
+	}
+
+	// Block until cluster has gone into update status and then completed
+	err = ng.WaitForClusterStatus(clusterStatusUpdateInProgress, waitForUpdateStatusTimeout)
+	if err != nil {
+		return fmt.Errorf("scale up error: %v", err)
+	}
+	err = ng.WaitForClusterStatus(clusterStatusUpdateComplete, waitForCompleteStatusTimout)
```

Comment:
Does this wait until new nodes actually boot and register in k8s? In general this method should just request VMs, not block until they boot up.

Reply:
The Nodegroup interface suggests that the scale up/down implementations should wait for the operation to complete.
https://github.com/kubernetes/autoscaler/blob/6b1a97c16001a98607424568faf7239b5c364d11/cluster-autoscaler/cloudprovider/cloud_provider.go#L92-L95

Reply:
It was designed with the assumption that underlying cloud provider has a 'desired size'. The method must wait until that 'desired size' is updated, not until VMs are actually created. 

CA takes 'upcoming' nodes into account in its simulations and it can trigger another scale-up while the previous nodes are still starting up if even more pending pods show up. You'll lose that ability if you block until VMs are created.

Reply:
> It was designed with the assumption that underlying cloud provider has a 'desired size'. The method must wait until that 'desired size' is updated, not until VMs are actually created.

Ok, that makes sense. The node_count in magnum is updated at the start of the scaling, so this would be the point where it should stop waiting.

Even so I have it set up so that any scales that are triggered while the cluster is an in updating status are ignored anyway. In my own experience letting it send another update to the cluster while it was still updating would always end up with the cluster in UPDATE_FAILED status.

Reply:
> CA takes 'upcoming' nodes into account in its simulations and it can trigger another scale-up while the previous nodes are still starting up if even more pending pods show up. You'll lose that ability if you block until VMs are created.

Do you have a pointer to the code that does the simulations? With our implementation all vms are identical, so the simulations could be made, but, another scale up will not be possible at the moment (this could change in the future)

Reply:
> Do you have a pointer to the code that does the simulations?

Half of CA codebase is dedicated to running simulations :) If you ask a specific question I can point you to relevant code, but just dealing with upcoming nodes would be hundreds LOC spread between various files in clusterstate/, core/ and expander/.

Reply:
> It was designed with the assumption that underlying cloud provider has a 'desired size'. The method must wait until that 'desired size' is updated, not until VMs are actually created.

Magnum and the underlying Heat stack don't have the notion of "Desired Size" being decoupled from a scaling event.  If the node count parameter changes on the stack, that immediately triggers the stack's size update policy.  There is also no imperative method to add or remove nodes with the Heat::ResourceGroup that the Magnum clusters are built on (there no "add X nodes" method for the ResourceGroup).  I think this situation can be made more consistent with how the other cloud providers work by keeping track of the "target size" as scaling events are processed by the cloudprovider code.  This would allow the openstack autoscaler code to be able to issue concurrent updates, as long as the previous "target size" was calculated correctly from the previous update (the mutex should ensure that the new "target size" is set before another thread evaluates a scaling operation)

> With our implementation all vms are identical, so the simulations could be made, but, another scale up will not be possible at the moment (this could change in the future)

From my tests, the Heat stack is able to handle multiple scale up requests before one finishes, as long as they know the true reference from the "delta" parameter.  I think this could be improved by updating the ng.targetSize as updates come in to the cloudprovider code.